paasta_tools/instance/kubernetes.py

```diff
@@ -407,6 +407,96 @@ def filter_actually_running_replicasets(
+    instance: str,
+    verbose: int,
+    include_smartstack: bool,
+    include_envoy: bool,
+    instance_type: str,
+    settings: Any,
+):
+    status: Dict[str, Any] = {}
+    config_loader = LONG_RUNNING_INSTANCE_TYPE_HANDLERS[instance_type].loader
+    job_config = config_loader(
+        service=service,
+        instance=instance,
+        cluster=settings.cluster,
+        soa_dir=settings.soa_dir,
+        load_deployments=True,
+    )
+    kube_client = settings.kubernetes_client
+    if kube_client is None:
+        return status
+
+    app_name = job_config.get_sanitised_deployment_name()
+    pod_list = kubernetes_tools.pods_for_service_instance(
+        service=job_config.service,
+        instance=job_config.instance,
+        kube_client=kube_client,
+        namespace=job_config.get_kubernetes_namespace(),
+    )
+    pods_by_replicaset = get_pods_by_replicaset(pod_list)
+    replicaset_list = kubernetes_tools.replicasets_for_service_instance(
+        service=job_config.service,
+        instance=job_config.instance,
+        kube_client=kube_client,
+        namespace=job_config.get_kubernetes_namespace(),
+    )
+    # For the purpose of active_shas/app_count, don't count replicasets that
+    # are at 0/0.
+    actually_running_replicasets = filter_actually_running_replicasets(replicaset_list)
+
+    desired_state = job_config.get_desired_state()
+
+    status["app_name"] = app_name
+    status["desired_state"] = desired_state
+    status["desired_instances"] = (
+        job_config.get_instances() if desired_state != "stop" else 0
+    )
+    status["bounce_method"] = job_config.get_bounce_method()
+    status["replicasets"] = [
+        get_replicaset_status(
+            replicaset, pods_by_replicaset.get(replicaset.metadata.name),
+        )
+        for replicaset in actually_running_replicasets
+    ]
+    return status
+
+
+def get_pods_by_replicaset(pods: Sequence[V1Pod]) -> Dict[str, List[V1Pod]]:
+    # TODO: should we make sure that we get every pod...?
+    pods_by_replicaset: DefaultDict[str, List[V1Pod]] = defaultdict(list)
+    for pod in pods:
+        for owner_reference in pod.metadata.owner_references:
+            if owner_reference.kind == "ReplicaSet":
+                pods_by_replicaset[owner_reference.name].append(pod)
+
+    return pods_by_replicaset
+
+
+def get_replicaset_status(
+    replicaset: V1ReplicaSet, pods: Sequence[V1Pod],
+) -> Dict[str, Any]:
+    return {
+        "name": replicaset.metadata.name,
+        "replicas": replicaset.spec.replicas,
+        "ready_replicas": ready_replicas_from_replicaset(replicaset),
+        "create_timestamp": replicaset.metadata.creation_timestamp.timestamp(),
+        "git_sha": replicaset.metadata.labels.get("paasta.yelp.com/git_sha"),
+        "config_sha": replicaset.metadata.labels.get("paasta.yelp.com/config_sha"),
+        "pods": [get_pod_status(pod) for pod in pods],
+    }
+
+
+def get_pod_status(pod: V1Pod) -> Dict[str, Any]:
```

Comment:
super-nit: should we make a `PodStatus` `TypedDict` so that we can replace this `Any`?

Reply:
or a `NamedTuple`/`dataclass`, I guess :p

Reply:
Yeah that would be nice but I guess I don't have time before I leave on PTO :-/